ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java

```diff
@@ -303,6 +305,90 @@ public void onMatch(RelOptRuleCall call) {
     }
   }
 
+  /**
+   * Determines weather the give grouping is unique.
+   *
+   * Consider a join which might produce non-unique rows; but later the results are aggregated again.
+   * This method determines if there are sufficient columns in the grouping which have been present previously as unique column(s).
+   */
+  private boolean isGroupingUnique(RelNode input, ImmutableBitSet groups) {
+    if (groups.isEmpty()) {
+      return false;
+    }
+    RelMetadataQuery mq = input.getCluster().getMetadataQuery();
+    Set<ImmutableBitSet> uKeys = mq.getUniqueKeys(input);
```

Comment:
If the purpose of this method is to determine that given a set of columns are unique or not you can use `areColumnsUnique` as @jcamachor  suggested.

Reply:
yes; I've explored using `areColumnsUnique` because it matches the usecase here - however for some tests it emitted some NPEs so I've gone back to the `getUniqueKeys` approach
I'll file a jira for `areColumnsUnique` when I know what's wrong with it..